exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DynamicResourceManager.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.work.foreman.rm;
+
+import org.apache.drill.exec.ExecConstants;
+import org.apache.drill.exec.ops.QueryContext;
+import org.apache.drill.exec.server.DrillbitContext;
+import org.apache.drill.exec.server.options.SystemOptionManager;
+import org.apache.drill.exec.work.foreman.Foreman;
+
+/**
+ * Wrapper around the default and/or distributed resource managers
+ * to allow dynamically enabling and disabling queueing.
+ */
+
+public class DynamicResourceManager implements ResourceManager {
+
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DynamicResourceManager.class);
+
+  private final DrillbitContext context;
+  private ResourceManager defaultRm;
+  private ResourceManager queueingRm;
+  private ResourceManager activeRm;
+  public long lastUpdateTime;
+  public int recheckDelayMs = 5000;
+
+  public DynamicResourceManager(final DrillbitContext context) {
+    this.context = context;
+    refreshRM();
+  }
+
+  public synchronized ResourceManager activeRM() {
+    refreshRM();
+    return activeRm;
+  }
+
+  @Override
+  public long memoryPerNode() {
+    return activeRm.memoryPerNode();
+  }
+
+  @Override
+  public int cpusPerNode() {
+    return activeRm.cpusPerNode();
+  }
+
+  @Override
+  public synchronized QueryResourceAllocator newQueryPlanner(QueryContext queryContext) {
```

Comment:
this synchronization will be a problem with high concurrency. Can we reduce the number of synchronizations we are doing per query ? Seems like you are doing when newQueryPlanner is called and newExecRM is called. Is one enough for both ?

Reply:
The refresh is done periodically, not every time that `refreshRM()` is called.

The problem is, we don't know if we are the thread that will cause the periodic refresh, or if the thread right after us will do it. If another thread does it, we might end up building a plan based on variables that change.

This an probably be rewritten in a functional style: return an object with current settings, and create a new object with new settings when the refresh occurs. Synchronization is still needed unless Java pointer assignment is thread-safe; I'd need to check.

In any event, this whole mechanism is more for larger queries. If we can run at very high concurrency, then each query uses very little resources and throttling becomes less necessary.